cloud/linode/instances.go

```diff
@@ -38,11 +85,31 @@ func (nc *nodeCache) refreshInstances(ctx context.Context, client client.Client)
 	if err != nil {
 		return err
 	}
-	nc.nodes = make(map[int]*linodego.Instance)
+	nc.nodes = make(map[int]*linodego.Instance, len(instances))
+	nc.ips = make(map[int][]nodeIP, len(instances))
+	wg := sync.WaitGroup{}
 	for _, instance := range instances {
-		instance := instance
-		nc.nodes[instance.ID] = &instance
+		wg.Add(1)
+		go func(ctx context.Context, instance linodego.Instance) {
+			defer wg.Done()
+
+			addresses, err := nc.getInstanceIPv4Addresses(ctx, instance.ID, client)
+			if err != nil {
+				klog.Errorf("Failed fetching ip addresses for instance id %d. Error: %s", instance.ID, err.Error())
+				nc.mtx.Lock()
+				defer nc.mtx.Unlock()
+				nc.nodes[instance.ID] = &instance
+				return
+			}
+			// take lock on map so that concurrent writes are safe
+			nc.mtx.Lock()
+			defer nc.mtx.Unlock()
+			nc.nodes[instance.ID] = &instance
```

Comment:
As discussed, since I am passing the instance by value, we should be good here:
```
go func(ctx context.Context, instance linodego.Instance) {
    ...
}(ctx, instance)
```